### ISSUE TITLE
Mipmapping

### ISSUE THREAD

**Issue by comnom** [endless-sky/endless-sky#3930]
I've done a test implementation of mipmapping to combat aliasing at the lowest zoom levels. Test branch here https://github.com/endless-sky/endless-sky/compare/master...comnom:mipmapping .

It's been a few years since there's been any real concerns for cards with extremely low vram (ie less than 256) so I figured the time was ripe to open a discussion for mipmapping.

Using the above implementation I've taken some screenshots to get things started. I focused on beams because they tend to be the worst offender's of aliasing at low zoom levels. I tried to match angles as best I could, but I recommend building and  trying it yourself, as aliasing is highly variable based on angle. 

Top is no mipmapping, bottom is with it (not vanilla, sorry didn't have a save handy). All images at 25% zoom:
<details>

![exampleonenomipmap](https://user-images.githubusercontent.com/22107649/45265939-1f7ce780-b421-11e8-9c1e-cc53e4514c46.png)

![exampleonemipmap](https://user-images.githubusercontent.com/22107649/45265943-3a4f5c00-b421-11e8-86f3-870a8856dd59.png)

![exampletwonomipmap](https://user-images.githubusercontent.com/22107649/45265942-302d5d80-b421-11e8-9803-c6ca1a386ab9.png)

![exampletwomipmap](https://user-images.githubusercontent.com/22107649/45265949-46d3b480-b421-11e8-855c-44c2e2b84d7f.png)
</details>

A couple notes
* Beam brightness is affected. The extra filtering seems to negatively impact the blend modes.
* Mipmapping does not eliminate all aliasing effects (but does diminish them).
* The "legacy mipmapping" preference you see in the code is to work around a bug in my radeon 3870 drivers that causes vram to increase by ~10x if I use `glGenerateMipmap(GL_TEXTURE_2D_ARRAY);`. Avior/Bladewood tested the same implementation and did not have a huge spike in vram. (he got around the 1.3x it should be on intel hd graphics). I should have an opportunity to test on a more modern amd card soon (an r9 280x).

I dont have any good profiling tools for windows to give any solid input there, but load times are impacted, and vram naturally goes up. This is definitely a try it before you buy it feature, so I encourage others to build it and play around.

**Comment by MCOfficer** [comment]
i am by no means expert in this matter, but FXAA seems like a much better performing alternative - at least after reading into it for a few minutes.

**Comment by comnom** [comment]
>FXAA seems like a much better performing alternative

I'd need profiling tools to say for sure (my card is powerful enough to not notice any difference with or without mipmapping or multisampling), but I did run a test at 8x multisampling (and no mipmapping).

I couldn't see any difference at all with multisampling enabled:
![msaa](https://user-images.githubusercontent.com/22107649/45270186-68a45a00-b468-11e8-8410-58985e354423.png)

Vram usuage was also higher with multisampling than mipmapping, my test had me at ~110MB (no msaa, no mipmapping), ~135MB (mipmapping), and ~200MB (8x msaa).

The book I've got doesn't go into too much detail about when/where/why for multisampling, so I'm not sure if we'd get better results by setting up an FBO for it rather than using one owned by the windowing system. (I am sure that's way more complex.)
